### PR TITLE
oneAPI kmeans Lloyd's algo impl

### PR DESCRIPTION
**Pull request by mchernov-intel**
- kmeans_init 
- kmeans train & infer

### CODE REVIEW

**Comment by mchernov-intel** [comment]
@PetrovKP @rlnx  please, review

File: examples/oneapi/cpp/makefile_lnx [hunk 1]
```diff
@@ -147,7 +147,7 @@ sointel64:
 _make_ex: $(RES)
 
 vpath
-vpath %.cpp $(addprefix ./source/,pca svm)
+vpath %.cpp $(addprefix ./source/,pca kmeans kmeans_init svm)
```
Comment (PetrovKP) [inline examples/oneapi/cpp/makefile_lnx:150]: I think it's better to add alphabetically

```suggestion
vpath %.cpp $(addprefix ./source/, kmeans kmeans_init pca svm)
```

File: examples/oneapi/cpp/source/kmeans/kmeans_dense_batch.cpp [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    const float data[] = { 1.f,  2.f, 3.f,  4.f,  5.f,  6.f,  -5.f,
+                           1.f, -1.f, 0.f,  3.f, 1.f, 2.f,  3.f,
+                           4.f,  5.f,  6.f,  1.f,  0.f,  0.f, 0.f,
+                           1.f,  2.f, 5.f, 2.f,  9.f, 3.f,  2.f,
+                           -4.f, 3.f,  0.f,  4.f, 2.f,  7.f,  5.f,
+                           4.f, 2.f,  0.f, -4.f, 0.f,  3.f,  -8.f,
+                           2.f,  5.f, 5.f,  -6.f, 3.f, 0.f, -9.f,
+                           3.f, 1.f,  -3.f, 3.f,  5.f,  1.f,  7.f };
+
+    const float initial_centroids[] = { 1.f,  2.f, 3.f,  4.f,  5.f,  6.f,  -5.f,
```
Comment (PetrovKP) [inline examples/oneapi/cpp/source/kmeans/kmeans_dense_batch.cpp:40]: Use algorithm: kmeans_init

File: examples/oneapi/cpp/source/kmeans/kmeans_dense_batch.cpp [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+                                        1.f,  2.f, 5.f, 2.f,  9.f, 3.f,  2.f };
+
+    const auto data_table = dal::homogen_table{ row_count, column_count, data };
+    const auto initial_centroids_table = dal::homogen_table{ 1, column_count, initial_centroids };
+
+    const auto kmeans_desc = dal::kmeans::descriptor<>()
```
Comment (PetrovKP) [inline examples/oneapi/cpp/source/kmeans/kmeans_dense_batch.cpp:46]:
```suggestion
    const auto kmeans_desc = dal::kmeans::descriptor()
```

File: examples/oneapi/cpp/source/kmeans/kmeans_dense_batch.cpp [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+                                 .set_cluster_count(nClusters)
+                                 .set_max_iteration_count(nIterations)
+                                 .set_accuracy_threshold(1.f);
```
Comment (PetrovKP) [inline examples/oneapi/cpp/source/kmeans/kmeans_dense_batch.cpp:49]: `accuracy_threshold` has double type
```suggestion
                                 .set_accuracy_threshold(1.0);
```

File: examples/oneapi/cpp/source/kmeans/kmeans_dense_batch.cpp [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+                           3.f, 1.f,  -3.f, 3.f,  5.f,  1.f,  7.f };
+
+    const float initial_centroids[] = { 1.f,  2.f, 3.f,  4.f,  5.f,  6.f,  -5.f,
+                                        1.f,  2.f, 5.f, 2.f,  9.f, 3.f,  2.f };
+
+    const auto data_table = dal::homogen_table{ row_count, column_count, data };
+    const auto initial_centroids_table = dal::homogen_table{ 1, column_count, initial_centroids };
+
+    const auto kmeans_desc = dal::kmeans::descriptor<>()
+                                 .set_cluster_count(nClusters)
```
Comment (PetrovKP) [inline examples/oneapi/cpp/source/kmeans/kmeans_dense_batch.cpp:47]: why make in a separate variable?
```suggestion
                                 .set_cluster_count(2)
```

File: examples/oneapi/cpp/source/kmeans/kmeans_dense_batch.cpp [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+                                 .set_max_iteration_count(nIterations)
```
Comment (PetrovKP) [inline examples/oneapi/cpp/source/kmeans/kmeans_dense_batch.cpp:48]:
```suggestion
                                 .set_max_iteration_count(5)
```

File: examples/oneapi/cpp/source/kmeans/kmeans_dense_batch.cpp (hunk 4 above)
Comment (PetrovKP) [reply examples/oneapi/cpp/source/kmeans/kmeans_dense_batch.cpp:49]: Very big threshold. Why did you choose this? In no case do not leave it for an example - it is very terrible from the point of view of mathematics. As a rule, it is set close to zero.

File: examples/oneapi/cpp/source/kmeans/kmeans_dense_batch.cpp [hunk 7]
```diff
@@ -0,0 +1,69 @@
+/*******************************************************************************
```
Comment (PetrovKP) [inline examples/oneapi/cpp/source/kmeans/kmeans_dense_batch.cpp:1]: How correctly does the example solve the problem of finding clusters. Did you check? It looks very suspicious.

Comment (PetrovKP) [inline examples/oneapi/cpp/source/kmeans/kmeans_dense_batch.cpp:1]: Rename example `kmeans_dense_batch.cpp` -> `kmeans_lloyd_dense_batch.cpp`. Algorithm - [Type of task (optional)] - Method - Input format - Mode

File: examples/oneapi/cpp/source/kmeans/kmeans_dense_batch.cpp [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include "oneapi/dal/algo/kmeans.hpp"
+#include "oneapi/dal/data/accessor.hpp"
```
Comment (PetrovKP) [inline examples/oneapi/cpp/source/kmeans/kmeans_dense_batch.cpp:18]:
```suggestion
```

File: examples/oneapi/cpp/source/kmeans/kmeans_dense_batch.cpp [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+#include "oneapi/dal/data/table.hpp"
```
Comment (PetrovKP) [inline examples/oneapi/cpp/source/kmeans/kmeans_dense_batch.cpp:19]: excessive
```suggestion
```

File: examples/oneapi/cpp/source/kmeans/kmeans_dense_batch.cpp [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+                                 .set_cluster_count(nClusters)
+                                 .set_max_iteration_count(nIterations)
+                                 .set_accuracy_threshold(1.f);
+
+    const auto result_train = dal::train(kmeans_desc, data_table, initial_centroids_table);
+
+    std::cout << "Iteration count: " << result_train.get_iteration_count() << std::endl;
+    std::cout << "Objective function value: " << result_train.get_objective_function_value() << std::endl;
+    std::cout << "Lables:" << std::endl << result_train.get_labels() << std::endl;
+    std::cout << "Centroids:" << std::endl << result_train.get_model().get_centroids() << std::endl;
```
Comment (PetrovKP) [inline examples/oneapi/cpp/source/kmeans/kmeans_dense_batch.cpp:56]: result_train should have method `get_centroids`. use him.
```suggestion
    std::cout << "Centroids:" << std::endl << result_train.get_centroids() << std::endl;
```

File: examples/oneapi/cpp/source/kmeans_init/kmeans_init_default_dense.cpp [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+                           4.f, 2.f,  0.f, -4.f, 0.f,  3.f,  -8.f,
+                           2.f,  5.f, 5.f,  -6.f, 3.f, 0.f, -9.f,
+                           3.f, 1.f,  -3.f, 3.f,  5.f,  1.f,  7.f };
+
+    const auto data_table = dal::homogen_table{ row_count, column_count, data };
+
+    const auto kmeans_init_desc = dal::kmeans_init::descriptor<>()
+        .set_cluster_count(nClusters);
+
+    const auto result = dal::train(kmeans_init_desc, data_table);
```
Comment (PetrovKP) [inline examples/oneapi/cpp/source/kmeans_init/kmeans_init_default_dense.cpp:60]:
```suggestion
    const auto result = dal::compute(kmeans_init_desc, data_table);
```

File: examples/oneapi/cpp/source/kmeans_init/kmeans_init_default_dense.cpp [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+                           4.f,  5.f,  6.f,  1.f,  0.f,  0.f, 0.f,
+                           1.f,  2.f, 5.f, 2.f,  9.f, 3.f,  2.f,
+                           -4.f, 3.f,  0.f,  4.f, 2.f,  7.f,  5.f,
+                           4.f, 2.f,  0.f, -4.f, 0.f,  3.f,  -8.f,
+                           2.f,  5.f, 5.f,  -6.f, 3.f, 0.f, -9.f,
+                           3.f, 1.f,  -3.f, 3.f,  5.f,  1.f,  7.f };
+
+    const auto data_table = dal::homogen_table{ row_count, column_count, data };
+
+    const auto kmeans_init_desc = dal::kmeans_init::descriptor<>()
```
Comment (PetrovKP) [inline examples/oneapi/cpp/source/kmeans_init/kmeans_init_default_dense.cpp:57]:
```suggestion
    const auto kmeans_init_desc = dal::kmeans_init::descriptor()
```

File: examples/oneapi/cpp/source/kmeans_init/kmeans_init_default_dense.cpp [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+#include <iomanip>
+#include <iostream>
+
+#include "oneapi/dal/algo/kmeans_init.hpp"
+#include "oneapi/dal/data/accessor.hpp"
+#include "oneapi/dal/data/table.hpp"
+
+using namespace oneapi;
+
+std::ostream& operator<<(std::ostream& stream, const dal::table& table) {
```
Comment (PetrovKP) [inline examples/oneapi/cpp/source/kmeans_init/kmeans_init_default_dense.cpp:26]: Use utils.

File: makefile.lst [hunk 14]
```diff
@@ -225,11 +225,15 @@ CORE.SERVICES       := compression            \
...
 # Dependencies between oneAPI and core (CPU-only) algorithms
 ONEAPI.ALGOS.pca := CORE.pca
+ONEAPI.ALGOS.kmeans := CORE.kmeans
+ONEAPI.ALGOS.kmeans_init := CORE.kmeans
 ONEAPI.ALGOS.svm := CORE.svm
 
 # List of algorithms in oneAPI part
 ONEAPI.ALGOS := \
     pca         \
+    kmeans      \
```
Comment (PetrovKP) [inline makefile.lst:235]: I think it's better to add alphabetically

File: cpp/oneapi/dal/algo/kmeans_init/backend/cpu/train_kernel_default_dense.cpp [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+        interop::convert_to_daal_homogen_table(arr_centroids, cluster_count, column_count);
+    const size_t nr                            = 1;
+    daal::data_management::NumericTable* r[nr] = { daal_centroids.get() };
+
+    interop::call_daal_kernel<Float, daal_kmeans_init_defaultDense_kernel_t>(ctx,
+                                                                             na,
+                                                                             a,
+                                                                             nr,
+                                                                             r,
+                                                                             &par,
```
Comment (PetrovKP) [inline cpp/oneapi/dal/algo/kmeans_init/backend/cpu/train_kernel_default_dense.cpp:62]: `na`, `a`, `nr`, `r` are these really good names?

File: cpp/oneapi/dal/algo/kmeans_init/train_types.cpp [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+*******************************************************************************/
+
+#include "oneapi/dal/algo/kmeans_init/train_types.hpp"
+#include "oneapi/dal/detail/common.hpp"
+
+namespace oneapi::dal::kmeans_init {
+
+class detail::train_input_impl : public base {
+public:
+    train_input_impl(const table& data) : data(data) {}
```
[the remark this answers is not in the document]
Comment (mchernov-intel) [reply cpp/oneapi/dal/algo/kmeans_init/train_types.cpp:24]: decided to postpone

File: examples/oneapi/cpp/source/kmeans_init/kmeans_init_default_dense.cpp (hunk 12 above)
Comment (mchernov-intel) [reply examples/oneapi/cpp/source/kmeans_init/kmeans_init_default_dense.cpp:42]: @rlnx what do you think?

File: cpp/oneapi/dal/algo/kmeans_init/common.hpp [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+
+namespace oneapi::dal::kmeans_init {
+
+namespace detail {
+struct tag {};
+class descriptor_impl;
+} // namespace detail
+
+namespace method {
+struct dense {};
```
[the remark this answers is not in the document]
Comment (mchernov-intel) [reply cpp/oneapi/dal/algo/kmeans_init/common.hpp:30]: decided to postpone

**Comment by mchernov-intel** [comment]
moved to #700

**Pull request closed without merging**